### PR TITLE
8298088: RISC-V: Make Address a discriminated union internally

### PR DESCRIPTION
**Pull request by RealFYang**
The RISC-V port has the same issue as: https://bugs.openjdk.org/browse/JDK-8297830
It looks to me that the fix for the AArch64 port is a nice refactoring work.
This fixes this issue for the RISC-V port with a similar approach.

Testing:
  Tier1 tested with release build on linux-riscv64 unmatched board.
  Run non-trivial benchmark workloads with fastdebug builds.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298088](https://bugs.openjdk.org/browse/JDK-8298088): RISC-V: Make Address a discriminated union internally

### Reviewers
 * [Feilong Jiang](https://openjdk.org/census#fjiang) (@feilongjiang - Author) ⚠️ Review applies to [ab1b8401](https://git.openjdk.org/jdk/pull/11505/files/ab1b84019f2278e4c1abe65e784e80e5431eae77)
 * [Yadong Wang](https://openjdk.org/census#yadongwang) (@yadongw - Author)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11505/head:pull/11505` \
`$ git checkout pull/11505`

Update a local copy of the PR: \
`$ git checkout pull/11505` \
`$ git pull https://git.openjdk.org/jdk pull/11505/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11505`

View PR using the GUI difftool: \
`$ git pr show -t 11505`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11505.diff">https://git.openjdk.org/jdk/pull/11505.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/assembler_riscv.hpp [hunk 1]
```diff
@@ -31,6 +31,11 @@
 #include "assembler_riscv.inline.hpp"
 #include "metaprogramming/enableIf.hpp"
```
Comment (feilongjiang) [inline src/hotspot/cpu/riscv/assembler_riscv.hpp:33]: I think this blank line can be removed.

**Review by feilongjiang — APPROVED** [review]
Nice cleanup! With one comment:

File: src/hotspot/cpu/riscv/assembler_riscv.hpp (hunk 1 above)
Comment (RealFYang) [reply src/hotspot/cpu/riscv/assembler_riscv.hpp]: Fixed. Thanks.

**Review by yadongw — APPROVED** [review]
lgtm

**Comment by RealFYang** [comment]
@feilongjiang @yadongw : Thanks for the review! @shipilev : Want to take a look?

**Review by shipilev — COMMENTED** [review]
AArch64 patch (https://github.com/openjdk/jdk/commit/5a5ced3a900a81fd0b0757017f4138ce97e2521e) also has modifications that rewrite `_offset` -> `offset()`, `_base` -> `base()`, `_index` -> `index()`, etc. Does RISC-V code has the relevant uses too?

**Comment by RealFYang** [comment]
> AArch64 patch ([5a5ced3](https://github.com/openjdk/jdk/commit/5a5ced3a900a81fd0b0757017f4138ce97e2521e)) also has modifications that rewrite `_offset` -> `offset()`, `_base` -> `base()`, `_index` -> `index()`, etc. Does RISC-V code has the relevant uses too?

Hi, I double-checked files assembler_riscv.hpp/cpp and macroAssembler_riscv.hpp/cpp. And I don't see any occurrence of direct uses for those private data members.

**Review by shipilev — APPROVED** [review]
Okay then!

**Comment by RealFYang** [comment]
> Okay then!

Thanks for the review. Let's
/integrate

**Pull request closed without merging**